shared/elements/gaia_grid/js/grid_dragdrop.js

```diff
@@ -489,16 +489,35 @@
         }
 
         // Cancel rearrangement if it would have no effect.
-        if (this.gridView.items[foundIndex].element.classList.
-            contains('invalid-drop')) {
-          if (insertDividerAtTop) {
-            this.container.classList.remove('hover-over-top');
+        var redundantRearrange = false;
+        if (insertDividerAtTop) {
+          if (iconIsDivider) {
+            redundantRearrange = true;
+            for (var i = foundIndex - 1; i >= 0; i--) {
+              if (this.gridView.items[i].detail.type === 'divider') {
+                redundantRearrange = false;
+                break;
+              }
+            }
+          } else {
+            if (this.icon.detail.index === 0 &&
+                this.gridView.items[1].detail.type === 'placeholder') {
+              redundantRearrange = true;
+            }
           }
+        } else {
+          redundantRearrange = this.gridView.items[foundIndex].element.
+            classList.contains('invalid-drop')
+        }
+
+        if (redundantRearrange) {
           if (this.hoverItem) {
             this.hoverItem.element.classList.remove('hovered');
             this.hoverItem = null;
           }
           return;
+        } else if (insertDividerAtTop) {
+          this.container.classList.add('hover-over-top');
```

Comment:
I don't think there's a simplification of the if/else structure that would do away with variables altogether - I'll think about it some, but I'd rather have the consequences grouped together here rather than duplicated and spread over several nested if/elses, if it came down to it.